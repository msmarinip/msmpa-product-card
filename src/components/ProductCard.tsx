import React from 'react';
import { createContext, CSSProperties } from 'react';

import { useProduct } from '../hooks/useProduct';
import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';

import styles from '../styles/styles.module.css';



export const ProductContext= createContext({} as ProductContextProps);
//Extraigo el provider del Context que es mi proveedor de info y lo coloco donde lo necesito
//para que todos lo hijos tengan acceso al provider, es lo primero que coloco en el ProductCard
const { Provider } = ProductContext;

//el children también podría ser () => JSX.Element
export interface Props {
    className?: string;
    value?: number;
    product: Product;
    // children?: ReactElement | ReactElement[]; //puede ser un arreglo de varios children
    children: (args: ProductCardHandlers) => JSX.Element;
    style?: CSSProperties;
    onChange?: (args: onChangeArgs) => void;
    initialValues?:InitialValues;
}


export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: Props) => {
    
    const {counter, isMaxCountReached,maxCount, increaseBy, reset} = useProduct({ onChange, product, value, initialValues });
    return (
        // En el value envío la info que necesitan los children
        <Provider value={{
            counter,
            increaseBy,
            product,
            maxCount
        }}>
            <div style={style} className={ `${styles.productCard} ${ className }` }>
                {/* React no puede renderizar una fción no puedo poner acá () => children 
                pero si puede ejecutarla*/}
                { 
                    children({
                        count: counter,
                        isMaxCountReached,
                        maxCount,
                        // maxCount: initialValues?.count,
                        product,
                        increaseBy,
                        reset
                    }) 
                }

            </div>
        </Provider>
    )
}
