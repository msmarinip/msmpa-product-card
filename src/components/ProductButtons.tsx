import React from 'react';
import { useContext, CSSProperties, useCallback } from 'react';
import { ProductContext } from "./ProductCard";

import styles from '../styles/styles.module.css';

export interface Props {
    className?: string;
    style?: CSSProperties;
    
}

export const ProductButtons = ({ className, style }: Props) => {
// maxCount del productContext

    const {increaseBy, counter, maxCount} = useContext(ProductContext)
// isMaxReached = useCallback, las dependencias son [ counter, maxCount]
        //regresa true si counter === maxCount, false si es distinto

    const isMaxReached =  useCallback(
        // () => {
        //     if(!maxCount) return;
        //     if ( counter === maxCount) {
        //         return true;
        //     } else {
        //         return false
        //     }
            
        // },
        // [counter, maxCount],
        //Más elegante así
        () => !!maxCount && counter === maxCount,
        [counter, maxCount]
    )
    return (
        <div 
            className={ `${styles.buttonsContainer} ${className}` }
            style={ style }
        >
            <button
                className={ styles.buttonMinus }
                onClick={ () => increaseBy(-1) }> - </button>
            <div className={ styles.countLabel }> { counter } </div>
            <button
                className={ `${ styles.buttonAdd } ${ (isMaxReached() && styles.disabled)  }` }
                onClick={ () => increaseBy(1) }> + </button>
        </div>
    );
}
