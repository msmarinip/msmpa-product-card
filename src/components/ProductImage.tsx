import React from 'react';

import { useContext } from 'react';
import { ProductContext } from './ProductCard';

import styles from '../styles/styles.module.css';

import noImage from '../assets/no-image.jpg';

export interface Props {
    img?: string;
    className?: string; 
    style?: React.CSSProperties;
}

//un string vacÃ­o en un ternario es considerado que no tiene valor => puedo poner por defaul img=''
export const ProductImage = ({ img = '', className, style }: Props) => {

    const { product } = useContext(ProductContext)
    let imgToShow: string;

    if( img ) {
        imgToShow = img;
    } else if (product.img) {
        imgToShow = product.img;
    } else {
        imgToShow = noImage;
    }
    
    
    return (
        // <img className={ styles.productImg } src={img ? img : noImage } alt="Product img" /> 
        <img 
            src={imgToShow } 
            className={ `${styles.productImg} ${className}` } 
            style={ style }
            alt="Product img" /> 
    )
}
