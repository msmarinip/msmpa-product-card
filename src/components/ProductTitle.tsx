import React from 'react';
import { useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';


export interface Props {
    title?: string; 
    className?: string;
    style?: React.CSSProperties;
}
//Para hecerlo diferente de la img, desestructuro el title de la interface y va a ser
//al hacerlo de esta manera el tÃ­tulo es obligatorio
// export const ProductTitle = ({ title, className }: { title?: string, className?:string }) => {
    export const ProductTitle = ({ title, className, style }: Props) => {
    const { product } = useContext(ProductContext)
    
    return (
        <span 
            className={ `${styles.productDescription} ${className}` }
            style={ style }
        >{ title ? title : product.title } </span>
    )
}