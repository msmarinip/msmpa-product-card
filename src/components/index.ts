import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductCardHOCProps } from '../interfaces/interfaces';

import { ProductTitle } from './ProductTitle';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';

export { ProductTitle } from './ProductTitle';
export { ProductImage } from './ProductImage';
export { ProductButtons } from './ProductButtons';


//assign para expandir propiedades al ProductCardHOC
export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
})

export default ProductCard;