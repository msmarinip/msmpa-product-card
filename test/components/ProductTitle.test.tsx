import React from 'react'
import renderer from 'react-test-renderer'
import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../data/products';

product1
describe('ProductTitle', () => {
    test('debe mostrar el componente correctamente con el tÃ­tulo personalizado', () => {
        
        const wrapper = renderer.create(
            <ProductTitle title="Custom Product" className="custom-class" style={{color: 'white'}}/>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
        // el test dio este error: Could not locate module ../styles/styles.module.css mapped as:     identity-obj-proxy.
        //tengo que instalar el identity-obj-proxy. en desarrollo
    })
    
    test('Debe mostrar el componente con el nombre del producto', () => {
        
        const wrapper = renderer.create(
            <ProductCard product={ product1 }>
                {
                    () => (
                        <ProductTitle />
                    )
                }
            </ProductCard> 
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
    })
    
})
