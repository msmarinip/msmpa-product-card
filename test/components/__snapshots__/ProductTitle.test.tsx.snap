// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle Debe mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug - Sin Imagen
     
  </span>
</div>
`;

exports[`ProductTitle debe mostrar el componente correctamente con el t√≠tulo personalizado 1`] = `
<span
  className="productDescription custom-class"
  style={
    Object {
      "color": "white",
    }
  }
>
  Custom Product
   
</span>
`;
